src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,734 @@
+  event.set_eventThread(tid);
+  event.commit();
+}
+
+void JfrCPUSamplerThread::post_run() {
+  this->NonJavaThread::post_run();
+  delete this;
+}
+
+static JfrCPUTimeThreadSampling* _instance = nullptr;
+
+JfrCPUTimeThreadSampling& JfrCPUTimeThreadSampling::instance() {
+  return *_instance;
+}
+
+JfrCPUTimeThreadSampling* JfrCPUTimeThreadSampling::create() {
+  assert(_instance == nullptr, "invariant");
+  _instance = new JfrCPUTimeThreadSampling();
+  return _instance;
+}
+
+void JfrCPUTimeThreadSampling::destroy() {
+  if (_instance != nullptr) {
+    delete _instance;
+    _instance = nullptr;
+  }
+}
+
+JfrCPUTimeThreadSampling::JfrCPUTimeThreadSampling() : _sampler(nullptr) {}
+
+JfrCPUTimeThreadSampling::~JfrCPUTimeThreadSampling() {
+  if (_sampler != nullptr) {
+    _sampler->disenroll();
+  }
+}
+
+void JfrCPUTimeThreadSampling::create_sampler(double rate, bool auto_adapt) {
+  assert(_sampler == nullptr, "invariant");
+  _sampler = new JfrCPUSamplerThread(rate, auto_adapt);
+  _sampler->start_thread();
+  _sampler->enroll();
+}
+
+void JfrCPUTimeThreadSampling::update_run_state(double rate, bool auto_adapt) {
+  if (rate != 0) {
+    if (_sampler == nullptr) {
+      create_sampler(rate, auto_adapt);
+    } else {
+      _sampler->set_rate(rate, auto_adapt);
+      _sampler->enroll();
+    }
+    return;
+  }
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate /* 0 */, auto_adapt);
+    _sampler->disenroll();
+  }
+}
+
+void JfrCPUTimeThreadSampling::set_rate(double rate, bool auto_adapt) {
+  assert(rate >= 0, "invariant");
+  if (_instance == nullptr) {
+    return;
+  }
+  instance().set_rate_value(rate, auto_adapt);
+}
+
+void JfrCPUTimeThreadSampling::set_rate_value(double rate, bool auto_adapt) {
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate, auto_adapt);
+  }
+  update_run_state(rate, auto_adapt);
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_create(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::trigger_async_processing_of_cpu_time_jfr_requests() {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->trigger_async_processing_of_cpu_time_jfr_requests();
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  if (Atomic::load_acquire(&_static_stop_signals)) {
+    return;
+  }
+  _instance->handle_timer_signal(info, context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(siginfo_t* info, void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (info->si_signo != SIGPROF) {
+    // not the signal we are interested in
```

Comment:
No, I meant checking `si_code`. `si_signo` will always be the right one.
And this check should come first, before any assertions.